workspace/__lib__/xod/graphics/triangle-outline-colored/patch.cpp

```diff
@@ -0,0 +1,51 @@
+
+// clang-format off
+{{#global}}
+#include <Triangle.h>
+{{/global}}
+// clang-format on
+
+struct State {
+    uint8_t mem[sizeof(TriangleOutline)];
+    TriangleOutline* triangleOutlineColored;
+    int16_t x0, y0, x1, y1, x2, y2;
+    XColor strokeColor;
+};
+
+// clang-format off
+{{ GENERATED_CODE }}
+// clang-format on
+
+void evaluate(Context ctx) {
+    auto state = getState(ctx);
+
+    auto gfx = getValue<input_GFX>(ctx);
+    state->strokeColor = getValue<input_C>(ctx);
+
+    int16_t x0 = (int16_t)getValue<input_X0>(ctx);
+    int16_t y0 = (int16_t)getValue<input_Y0>(ctx);
+    int16_t x1 = (int16_t)getValue<input_X1>(ctx);
+    int16_t y1 = (int16_t)getValue<input_Y1>(ctx);
+    int16_t x2 = (int16_t)getValue<input_X2>(ctx);
+    int16_t y2 = (int16_t)getValue<input_Y2>(ctx);
+
+    if (isSettingUp()) {
+        state->triangleOutlineColored = new (state->mem) TriangleOutline(gfx);
+    }
+
+    if (isSettingUp() || x0 != state->x0 || y0 != state->y0 || x1 != state->x1 || y1 != state->y1 || x2 != state->x2 || y2 != state->y2 || isInputDirty<input_C>(ctx)) {
```

Comment:
If anything changed — update state and emit updated triangle... looks like we can get rid of this condition (in each triangle node). Or I missed something?

Reply:
That is done to prevent node evaluate on number (float) changes.
Coorinates are integers only. For example, the change  `5.67` -> `5.68` should not trigger node evaluate because it is still `5`.